### ISSUE TITLE
cmd/yacc: move to subrepo

### ISSUE THREAD

**Issue by rsc** [golang/go#11229]
For Go 1.6 I intend to replace the use of cmd/yacc in the compiler with a hand-written recursive descent parser, as much a line-by-line translation of the current go.y as is possible. I've already done the conversion once, a month ago, although I'll have to redo it due to recent compiler changes.

After that's done, there will be no uses of cmd/yacc in the main repo. A recent mail to golang-dev points out that cmd/yacc has never shipped with gccgo. 

Perhaps for Go 1.6 we should think about moving cmd/yacc out of the main distribution and into a subrepo. It will be easier for people to 'go get' and stay up to date, and there will no longer be any compelling reason for keeping it in the main repo (if there ever was).

**Comment by rsc** [comment]
Didn't happen for Go 1.6.

**Comment by mdempsky** [comment]
Is this something to consider for Go 1.7?

**Comment by bradfitz** [comment]
Yes, let's do it.

**Comment by mdempsky** [comment]
Should cmd/go still install it into $GOTOOLDIR so "go tool yacc" continues to work?  Or do we demote it to a normal external command?  If a normal command, should we rename it "goyacc" or something to avoid conflicting with the standard yacc command?

Also, I'm assuming we want to move it to the x/tools subdirectory?

**Comment by bradfitz** [comment]
Demoting to regular as "goyacc" in x/tools SGTM.

**Comment by minux** [comment]
Have we announced this plan publicly? I think removing it without
prior notice will break a lot of //go:generate lines.

I propose that we make a copy of cmd/yacc to x/tools/cmd/goyacc
at the start of the Go 1.7 freeze, but we release cmd/yacc one more
time in Go 1.7 and call out in the release docs that Go 1.7 is the last
release containing `go tool yacc` and prompt users to switch to
x/tools/cmd/goyacc.

**Comment by minux** [comment]
Alternatively, we can make go tool yacc print out an error
message and prompt the user to use x/tools/cmd/goyacc
instead.

**Comment by cznic** [comment]
Please note that, for years, there is an existing [goyacc](https://godoc.org/github.com/cznic/goyacc) project. If the x/tools/goyacc will overwrite the goyacc binary, it will break things for users of, for example, [ql](https://github.com/cznic/ql), [cc](https://github.com/cznic/cc), [pl0](https://github.com/cznic/pl0), [gc(WIP)](https://github.com/cznic/gc), ...

**Comment by bradfitz** [comment]
> Please note that, for years, there is an existing goyacc project.

Maybe `yc`. :)

> Have we announced this plan publicly?

This bug has been open for a year. That doesn't mean everybody would see it, but not everybody always does.

> I think removing it without prior notice will break a lot of //go:generate lines.

That's okay.

1. we don't guarantee stability of tools. Only API.
2. this doesn't break building of programs or end users. It only breaks developers updating their programs. They're by definition already changing stuff, so they can change their go:generate line too.

**Comment by cznic** [comment]
> Maybe yc. :)

`yc` sound like a compiler to me. `yy` or `y` would be perhaps ok, but they're [both](https://godoc.org/github.com/cznic/yy) already [taken](https://godoc.org/github.com/cznic/y) ;-)

**Comment by bradfitz** [comment]
Maybe you should stop writing these tools and taking all the names. :-)

**Comment by mdempsky** [comment]
> `yc` sound like a compiler to me.

Nothing wrong with that: "yacc" is an acronym for "yet another compiler compiler" anyway.

**Comment by cznic** [comment]
> Nothing wrong with that: "yacc" is an acronym for "yet another compiler compiler" anyway.

Fair enough.

Nonetheless, dear Go team, if possible, please consider a different name than `goyacc`. If not possible/not preferred/whatever the reasons are, please state that decision early/ASAP to increase the time span in which I can pre-announce changing the name of `cznic/goyacc` to probably, @bradfitz credited, `yc`. The world will not end in the later case, I guess B-)

PS: Another name clash is the fork of cznic/goyacc the [TIDB project](https://github.com/pingcap/tidb/blob/d01d516105d2bc33ace72e2b310aab0263f7711a/Makefile#L66) is using.

**Comment by bradfitz** [comment]
@mdempsky, any interest in doing this for 1.7? It'd be nice to delete code from the main repo.

Actually, do we even need to move this? Can we just delete it and direct people towards an/the existing `goyacc` project that's actually maintained?

**Comment by mdempsky** [comment]
@cznic Is your goyacc tool backwards compatible with go tool yacc?

I know for cmd/compile, we added some new error message handling code to cmd/yacc, but it's still undocumented (#10888) so might not be a big deal to lose it.

**Comment by cznic** [comment]
tl;dr It's not.

It is mostly compatible, but not 100%. For plain yacc programs there should be no special problem in switching from go tool yacc to goyac. The improved error messages mechanism is not compatible at all. Go tool yacc uses non standard yacc extensions, goyacc uses an external xerrors file. Moreover, this mechanism in go tool yacc is more capable than that of goyacc. Also the produced error messages are different not only literally, bot also the approach is different. Goyacc produces the full LA set in the message, which is noisy, although nice when debugging grammars.

**Comment by rsc** [comment]
Too late for Go 1.7. Please focus on fixing things. This can only break things.

**Comment by rsc** [comment]
FWIW I think calling it goyacc is fine. We used the name before cznic did.

**Comment by cznic** [comment]
That's correct. FWIW, I started to use the name goyacc when you stopped to use it.

**Comment by rsc** [comment]
I think the world will not be confused by two goyaccs. :-)
